File: classic150/Solution0189.cpp
#include <vector>
#include <algorithm>
#include <numeric>
using namespace std;
// 给定一个整数数组 nums，将数组中的元素向右轮转 k 个位置，其中 k 是非负数。

// 示例 1:

// 输入: nums = [1,2,3,4,5,6,7], k = 3
// 输出: [5,6,7,1,2,3,4]
// 解释:
// 向右轮转 1 步: [7,1,2,3,4,5,6]
// 向右轮转 2 步: [6,7,1,2,3,4,5]
// 向右轮转 3 步: [5,6,7,1,2,3,4]
// 示例 2:

// 输入：nums = [-1,-100,3,99], k = 2
// 输出：[3,99,-1,-100]
// 解释:
// 向右轮转 1 步: [99,-1,-100,3]
// 向右轮转 2 步: [3,99,-1,-100]

// 提示：

// 1 <= nums.length <= 105
// -231 <= nums[i] <= 231 - 1
// 0 <= k <= 105

// 进阶：

// 尽可能想出更多的解决方案，至少有 三种 不同的方法可以解决这个问题。
// 你可以使用空间复杂度为 O(1) 的 原地 算法解决这个问题吗？
class Solution
{
public:
    void rotate(vector<int> &nums, int k)
    {
        k = k % nums.size();
        reverse(nums.begin(), nums.begin() + nums.size() - k); // 反转前n-k个元素
        reverse(nums.begin() + nums.size() - k, nums.end());   // 反转后k个元素
        reverse(nums.begin(), nums.end());                     // 反转整个数组
    }
};

class Solution
{
public:
    void rotate(vector<int> &nums, int k)
    {
        k = k % nums.size();
        int count = gcd(nums.size(), k); // 求最大公约数
        for (int i = 0; i < count; i++)
        {
            int current = i;
            int prev = nums[i];
            do
            {
                int next = (current + k) % nums.size();
                swap(nums[next], prev);
                current = next;
            } while (i != current);
        }
    }
};

File: anew/Solution0189.java
package anew;

// 给定一个整数数组 nums，将数组中的元素向右轮转 k 个位置，其中 k 是非负数。

public class Solution0189 {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        int count = gcd(n, k);
        for (int i = 0; i < count; i++) {
            int current = i;
            int currentValue = nums[current];
            do {
                int next = (current + k) % n;
                int temp = nums[next];
                nums[next] = currentValue;
                currentValue = temp;
                current = next;
            } while (current != i);
        }
    }

    private int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    public void rotate2(int[] nums, int k) {
        int[] temp = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            temp[(i + k) % nums.length] = nums[i];
        }
        System.arraycopy(temp, 0, nums, 0, nums.length);
    }

    public void rotate3(int[] nums, int k) {
        k = k % nums.length;
        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, nums.length - 1);
    }

    private void reverse(int[] nums, int left, int right) {
        while (left < right) {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        int[] nums = { 1, 2, 3, 4, 5, 6, 7 };
        int k = 3;
        new Solution0189().rotate(nums, k);
        for (int num : nums) {
            System.out.print(num + " ");
        }
        System.out.println();

        int[] nums2 = { 1, 2, 3, 4, 5, 6, 7 };
        new Solution0189().rotate2(nums2, k);
        for (int num : nums2) {
            System.out.print(num + " ");
        }
        System.out.println();

        int[] nums3 = { 1, 2, 3, 4, 5, 6, 7 };
        new Solution0189().rotate3(nums3, k);
        for (int num : nums3) {
            System.out.print(num + " ");
        }
    }
}
